src/workers/webhook-emitter.ts

```diff
@@ -0,0 +1,207 @@
+      );
+      return;
+    }
+
+    if (this.indexFilter.constructor.name === NeverMatch.name) {
+      this.log.info(
+        'WebhookEmitter not initialized. IndexFilter is set to NeverMatch.',
+      );
+      return;
+    }
+
+    this.log.info('WebhookEmitter initialized.');
+
+    await this.registerEventListeners();
+  }
+
+  public validateTargetServersUrls(): boolean {
+    // Check if all target servers URLs are http or https
+    return this.targetServersUrls.every((serverUrl) => {
+      try {
+        const url = new URL(serverUrl);
+        return url.protocol === 'http:' || url.protocol === 'https:';
+      } catch (error) {
+        return false;
+      }
+    });
+  }
+
+  public async shutdown(): Promise<void> {
+    // Remove specific listeners
+    for (const [event, listener] of this.listenerReferences) {
+      this.eventEmitter.removeListener(event, listener);
+    }
+    this.listenerReferences.clear();
+
+    await this.emissionQueue.kill();
+    this.log.info('WebhookEmitter shutdown completed.');
+  }
+
+  private async registerEventListeners(): Promise<void> {
+    this.log.debug('Registering WebhookEmitter listeners.');
+
+    for (const event of this.indexEventsToListenFor) {
+      const listener = async (data: WebhookEmissionData) => {
+        if (await this.indexFilter.match(data)) {
+          for (const targetServer of this.targetServersUrls) {
+            if (this.emissionQueue.length() < this.maxEmissionQueueSize) {
+              this.log.debug('Adding webhook to queue', {
+                event,
+                id: data.id,
+              });
+
+              this.emissionQueue.push({
+                targetServer,
+                eventWrapper: { event, data },
+              });
+            } else {
+              this.log.debug('Webhook queue is full. Skipping webhook:', {
+                event,
+                id: data.id,
+              });
+            }
+          }
+        }
+      };
+
+      this.listenerReferences.set(event, listener);
+      this.eventEmitter.on(event, listener);
+    }
+  }
+
+  public async emitWebhookToTargetServer(
+    details: WebhookEmissionDetails,
+  ): Promise<void> {
+    const { targetServer, eventWrapper } = details;
+    this.log.debug(
+      `Emitting webhook to ${targetServer} for ${eventWrapper.event}`,
+    );
+
+    try {
+      const response = await axios.post(targetServer, eventWrapper);
+
+      if (response.status >= 200 && response.status < 300) {
+        this.log.info(
+          `Webhook emitted successfully for: ${eventWrapper.data.id}`,
+        );
+      } else {
+        this.log.error(
+          `Failed to emit webhook. Status code: ${response.status}`,
+        );
+      }
+    } catch (error) {
+      if (axios.isAxiosError(error)) {
+        this.log.error('Error while emitting webhook:', error.message);
+      } else {
+        this.log.error('Unexpected error while emitting webhook:', error);
```

Comment:
I've been using this pattern for error logging: https://github.com/ar-io/ar-io-node/blob/develop/src/workers/transaction-offset-importer.ts#L84

It seems to work pretty well in most cases.